Firestore/Swift/Source/Codable/FirestoreDecoder.swift

```diff
@@ -0,0 +1,1055 @@
+    }
+    
+    func unbox(_ value: Any, as type: UInt16.Type) throws -> UInt16? {
+        guard !(value is NSNull) else { return nil }
+        
+        guard let number = value as? NSNumber, number !== kCFBooleanTrue, number !== kCFBooleanFalse else {
+            throw DecodingError._typeMismatch(at: self.codingPath, expectation: type, reality: value)
+        }
+        
+        let uint16 = number.uint16Value
+        guard NSNumber(value: uint16) == number else {
+            throw DecodingError.dataCorrupted(DecodingError.Context(codingPath: self.codingPath, debugDescription: "Parsed JSON number <\(number)> does not fit in \(type)."))
+        }
+        
+        return uint16
+    }
+    
+    func unbox(_ value: Any, as type: UInt32.Type) throws -> UInt32? {
+        guard !(value is NSNull) else { return nil }
+        
+        guard let number = value as? NSNumber, number !== kCFBooleanTrue, number !== kCFBooleanFalse else {
+            throw DecodingError._typeMismatch(at: self.codingPath, expectation: type, reality: value)
+        }
+        
+        let uint32 = number.uint32Value
+        guard NSNumber(value: uint32) == number else {
+            throw DecodingError.dataCorrupted(DecodingError.Context(codingPath: self.codingPath, debugDescription: "Parsed JSON number <\(number)> does not fit in \(type)."))
+        }
+        
+        return uint32
+    }
+    
+    func unbox(_ value: Any, as type: UInt64.Type) throws -> UInt64? {
+        guard !(value is NSNull) else { return nil }
+        
+        guard let number = value as? NSNumber, number !== kCFBooleanTrue, number !== kCFBooleanFalse else {
+            throw DecodingError._typeMismatch(at: self.codingPath, expectation: type, reality: value)
+        }
+        
+        let uint64 = number.uint64Value
+        guard NSNumber(value: uint64) == number else {
+            throw DecodingError.dataCorrupted(DecodingError.Context(codingPath: self.codingPath, debugDescription: "Parsed JSON number <\(number)> does not fit in \(type)."))
+        }
+        
+        return uint64
+    }
+    
+    func unbox(_ value: Any, as type: Float.Type) throws -> Float? {
+        guard !(value is NSNull) else { return nil }
+        
+        if let number = value as? NSNumber, number !== kCFBooleanTrue, number !== kCFBooleanFalse {
+            // We are willing to return a Float by losing precision:
+            // * If the original value was integral,
+            //   * and the integral value was > Float.greatestFiniteMagnitude, we will fail
+            //   * and the integral value was <= Float.greatestFiniteMagnitude, we are willing to lose precision past 2^24
+            // * If it was a Float, you will get back the precise value
+            // * If it was a Double or Decimal, you will get back the nearest approximation if it will fit
+            let double = number.doubleValue
+            guard abs(double) <= Double(Float.greatestFiniteMagnitude) else {
```

Comment:
+ @mikelehen What's the right way to respond to this based on our Android implementation?

It seems like CustomClassMapper just relies on "cast" operations (`Number#floatValue`): non-failing, saturating casts that would have e.g. positive values beyond ~3.4 x 10^38 saturate to Infinity when assigned to a model value containing a Float instead of a double. Isn't a similar response appropriate here?